sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala

```diff
@@ -66,49 +66,52 @@ private[sql] class HiveSessionCatalog(
       name: String,
       clazz: Class[_],
       input: Seq[Expression]): Expression = {
-
-    Try(super.makeFunctionExpression(name, clazz, input)).getOrElse {
-      var udfExpr: Option[Expression] = None
-      try {
-        // When we instantiate hive UDF wrapper class, we may throw exception if the input
-        // expressions don't satisfy the hive UDF, such as type mismatch, input number
-        // mismatch, etc. Here we catch the exception and throw AnalysisException instead.
-        if (classOf[UDF].isAssignableFrom(clazz)) {
-          udfExpr = Some(HiveSimpleUDF(name, new HiveFunctionWrapper(clazz.getName), input))
-          udfExpr.get.dataType // Force it to check input data types.
-        } else if (classOf[GenericUDF].isAssignableFrom(clazz)) {
-          udfExpr = Some(HiveGenericUDF(name, new HiveFunctionWrapper(clazz.getName), input))
-          udfExpr.get.dataType // Force it to check input data types.
-        } else if (classOf[AbstractGenericUDAFResolver].isAssignableFrom(clazz)) {
-          udfExpr = Some(HiveUDAFFunction(name, new HiveFunctionWrapper(clazz.getName), input))
-          udfExpr.get.dataType // Force it to check input data types.
-        } else if (classOf[UDAF].isAssignableFrom(clazz)) {
-          udfExpr = Some(HiveUDAFFunction(
-            name,
-            new HiveFunctionWrapper(clazz.getName),
-            input,
-            isUDAFBridgeRequired = true))
-          udfExpr.get.dataType // Force it to check input data types.
-        } else if (classOf[GenericUDTF].isAssignableFrom(clazz)) {
-          udfExpr = Some(HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName), input))
-          udfExpr.get.asInstanceOf[HiveGenericUDTF].elementSchema // Force it to check data types.
+    // Current thread context classloader may not be the one loaded the class. Need to switch
+    // context classloader to initialize instance properly.
+    Utils.withContextClassLoader(clazz.getClassLoader) {
+      Try(super.makeFunctionExpression(name, clazz, input)).getOrElse {
+        var udfExpr: Option[Expression] = None
+        try {
+          // When we instantiate hive UDF wrapper class, we may throw exception if the input
+          // expressions don't satisfy the hive UDF, such as type mismatch, input number
+          // mismatch, etc. Here we catch the exception and throw AnalysisException instead.
+          if (classOf[UDF].isAssignableFrom(clazz)) {
+            udfExpr = Some(HiveSimpleUDF(name, new HiveFunctionWrapper(clazz.getName), input))
+            udfExpr.get.dataType // Force it to check input data types.
```

Comment:
Found a potential problem: here we call `HiveSimpleUDF.dateType` (which is a lazy val), to force to load the class with the corrected class loader.

However, if the expression gets transformed later, which copies `HiveSimpleUDF`, then calling  `HiveSimpleUDF.dataType` will re-trigger the class loading, and at that time there is no guarantee that the corrected classloader is used.

I think we should materialize the loaded class in `HiveSimpleUDF`.

@HeartSaVioR can you take a look?

Reply:
Thanks for pinging me.

Could you please confirm my understanding? Actually my knowledge to resolve this issue came from debugging (like, reverse-engineering) so I'm not sure I get it 100%.

If my understanding is correct, this seems to be the simple reproducer - could you please confirm I understand correctly?

```
// uses classloader which loads clazz
val udf = HiveSimpleUDF(name, new HiveFunctionWrapper(clazz.getName), input)
udf.dataType
val newUdf = udf.makeCopy(Array.empty)
// change classloader which doesn't load clazz
newUdf.dataType
```

Reply:
yup, `HiveSimpleUDF` needs to load class when `dataType` is first called. So even if we load the class here in `HiveSessionCatalog`, but once `HiveSimpleUDF` is copied during transformation, it needs to load class again.

Reply:
I figured out above code doesn't give error - HiveFunctionWrapper stores `instance` which is copied in `makeCopy()` - so once the instance is created it doesn't seems to require changing classloader.

That said, below code gives error:

```
// uses classloader which loads clazz
val udf = HiveGenericUDTF(name, new HiveFunctionWrapper(clazz.getName), input)
// make sure HiveFunctionWrapper.createFunction is not called here

val newUdf = udf.makeCopy(udf.productIterator.map(_.asInstanceOf[AnyRef]).toArray)
// change classloader which doesn't load clazz
newUdf.dataType
``` 

we force call `.dataType` after creating HiveXXXUDF, so if my understanding is correct it won't be matter.

Could you please check whether my observation is correct, or please let me know if I'm missing something?

Reply:
The experimental UT code I used is below (added to SQLQuerySuite.scala) :

```
test("SPARK-26560 ...experimenting Wenchen's comment...") {
    // force to use Spark classloader as other test (even in other test suites) may change the
    // current thread's context classloader to jar classloader
    Utils.withContextClassLoader(Utils.getSparkClassLoader) {
      withUserDefinedFunction("udtf_count3" -> false) {
        val sparkClassLoader = Thread.currentThread().getContextClassLoader

        // This jar file should not be placed to the classpath; GenericUDTFCount3 is slightly
        // modified version of GenericUDTFCount2 in hive/contrib, which emits the count for
        // three times.
        val jarPath = "src/test/noclasspath/TestUDTF-spark-26560.jar"
        val jarURL = s"file://${System.getProperty("user.dir")}/$jarPath"

        val className = "org.apache.hadoop.hive.contrib.udtf.example.GenericUDTFCount3"
        sql(
          s"""
             |CREATE FUNCTION udtf_count3
             |AS '$className'
             |USING JAR '$jarURL'
          """.stripMargin)

        assert(Thread.currentThread().getContextClassLoader eq sparkClassLoader)

        // JAR will be loaded at first usage, and it will change the current thread's
        // context classloader to jar classloader in sharedState.
        // See SessionState.addJar for details.
        sql("SELECT udtf_count3(a) FROM (SELECT 1 AS a FROM src LIMIT 3) t")

        assert(Thread.currentThread().getContextClassLoader ne sparkClassLoader)
        assert(Thread.currentThread().getContextClassLoader eq
          spark.sqlContext.sharedState.jarClassLoader)

        // uses classloader which loads clazz
        val name = "default.udtf_count3"

        val input = Array(AttributeReference("a", IntegerType, nullable = false)())
        val udf = HiveGenericUDTF(name, new HiveFunctionWrapper(className), input)
        // FIXME: uncommenting below line will lead test passing
//        udf.dataType

        // Roll back to the original classloader and run query again. Without this line, the test
        // would pass, as thread's context classloader is changed to jar classloader. But thread
        // context classloader can be changed from others as well which would fail the query; one
        // example is spark-shell, which thread context classloader rolls back automatically. This
        // mimics the behavior of spark-shell.
        Thread.currentThread().setContextClassLoader(sparkClassLoader)

        // FIXME: doing this "within" the context classloader which loads the UDF class will
        //   lead test passing even we comment out udf.dataType
        val newUdf = udf.makeCopy(udf.productIterator.map(_.asInstanceOf[AnyRef]).toArray)

        newUdf.dataType
      }
    }
  }
```

Reply:
OK let me put my findings: If you look at `HiveFunctionWrapper.createFunction`, it says we don't cache the instance for Simple UDF
```
    def createFunction[UDFType <: AnyRef](): UDFType = {
      if (instance != null) {
        instance.asInstanceOf[UDFType]
      } else {
        val func = Utils.getContextOrSparkClassLoader
          .loadClass(functionClassName).newInstance.asInstanceOf[UDFType]
        if (!func.isInstanceOf[UDF]) {
          // We cache the function if it's no the Simple UDF,
          // as we always have to create new instance for Simple UDF
          instance = func
        }
        func
      }
    }
```
I don't know the history but I assume "we always have to create new instance for Simple UDF" is correct. I think what we can do is to cache the loaded `Class` as well as the instance.

Reply:
Oh OK. I missed the case we don't cache the function. Thanks for the pointer!
I'll try to reproduce the finding, and fix it without touching assumption.

Reply:
#28079